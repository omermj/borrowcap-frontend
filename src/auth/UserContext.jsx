import React from "react";

/** Context to provide currentUser thoughout the app */

const UserContext = React.createContext();

export default UserContext;
