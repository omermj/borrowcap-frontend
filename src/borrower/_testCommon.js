import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = () => {
  // Mock api call with active requests
  BorrowcapApi.getActiveRequestsByBorrowerId.mockResolvedValue([
    {
      id: 4,
      amtRequested: 10000,
      purpose: "Business",
      appOpenDate: "2021/01/01",
      interestRate: 0.05,
      term: 12,
      installmentAmt: 1000,
    },
    {
      id: 5,
      amtRequested: 20000,
      purpose: "Home",
      appOpenDate: "2021/02/01",
      interestRate: 0.07,
      term: 24,
      installmentAmt: 2000,
    },
  ]);

  BorrowcapApi.getApprovedRequestsByUserId.mockResolvedValue({
    borrower: [
      {
        id: 2,
        amtRequested: "13000",
        amtApproved: "12000",
        amtFunded: "6000",
        purpose: "Education",
        appOpenDate: "2023-01-01",
        appApprovedDate: "2023/01/05",
        fundingDeadline: "2023/02/05",
        interestRate: "0.07",
        term: 24,
        installmentAmt: "537.27",
        availableForFunding: false,
      },
    ],
  });
};

module.exports = { testUser, commonBeforeEach };
