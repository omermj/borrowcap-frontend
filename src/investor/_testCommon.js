import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getFundedLoansByUserId.mockResolvedValue({
    investor: [
      {
        id: 1,
        investorId: 3,
        amtInvested: "5000",
        amtFunded: "5000",
        fundedDate: "2023/01/01",
        interestRate: "0.05",
        term: 12,
        installmentAmt: "428.04",
        remainingBalance: "5000",
      },
    ],
  });
  await BorrowcapApi.getAvailableInvestments.mockResolvedValue([
    {
      id: 7,
      amtApproved: "8000",
      amtFunded: "0",
      purpose: "Medical",
      approvedDate: "2023/01/01",
      fundingDeadline: "2023/02/1",
      interestRate: "0.07",
      term: 24,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
