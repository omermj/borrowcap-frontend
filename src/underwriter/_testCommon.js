import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getLoanRequests.mockResolvedValue([
    {
      id: 2,
      borrowerId: 2,
      amtRequested: "10000",
      purpose: "Education",
      purposeId: 3,
      appOpenDate: "2023/01/01",
      interestRate: "0.07",
      term: 36,
      installmentAmt: "319.86",
    },
  ]);

  await BorrowcapApi.getApprovedRequests.mockResolvedValue([
    {
      id: 1,
      borrowerId: 2,
      amtRequested: "10000",
      amtApproved: "9000",
      amtFunded: "5000",
      purpose: "Car",
      purposeId: 2,
      appOpenDate: "2023/01/01",
      appApprovedDate: "2023/01/05",
      fundingDeadline: "2023/02/05",
      interestRate: "0.05",
      term: 12,
      installmentAmt: "428.04",
      availableForFunding: false,
      isFunded: true,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
