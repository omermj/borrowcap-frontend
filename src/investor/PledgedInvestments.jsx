import { useState, useEffect, useContext } from "react";
import TableComponent from "../common/TableComponent";
import TableHeader from "../common/Table/TableHeader";
import { formatCurrency, formatDate, formatPercent } from "../helpers/format";
import BorrowcapApi from "../api/api";
import UserContext from "../auth/UserContext";

/** Table showing all Active Investments made by the logged in user */

const PledgedInvestments = () => {
  const { currentUser } = useContext(UserContext);
  const [data, setData] = useState([]);

  // Define table headers with labels and formatters
  const headers = {
    id: { label: "ID", formatter: "none" },
    amtApproved: { label: "Approved Amount", formatter: formatCurrency },
    amtPledged: { label: "Pledged Amount", formatter: formatCurrency },
    approvedDate: { label: "Approved Date", formatter: formatDate },
    fundingDeadline: { label: "Funding Deadline", formatter: formatDate },
    interestRate: { label: "Interest Rate", formatter: formatPercent },
    term: { label: "Term", formatter: "none" },
  };

  // get active investments on initial render
  useEffect(() => {
    async function fetchPledgedInvestments() {
      const pledgedInvestments = await BorrowcapApi.getApprovedRequestsByUserId(
        currentUser.id
      );
      setData(pledgedInvestments.investor);
    }
    fetchPledgedInvestments();
  }, [currentUser.id]);

  return (
    <div className="border mb-4">
      <TableHeader text={"Pledged Investments"} />
      {!data.length ? (
        <div className=" mb-2 fst-italic fw-light">
          {" "}
          <small>No pledged investments.</small>
        </div>
      ) : (
        <TableComponent headers={headers} tableData={data} />
      )}
    </div>
  );
};

export default PledgedInvestments;
