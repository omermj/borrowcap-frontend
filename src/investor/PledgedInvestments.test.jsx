import { render, waitFor, screen, act } from "@testing-library/react";
import PledgedInvestments from "./PledgedInvestments.jsx";
import { expect, it } from "vitest";
import UserContext from "../auth/UserContext.jsx";
import { commonBeforeEach, testUser } from "./_testCommon.js";
import BorrowcapApi from "../api/api.js";

// Mock api calls
vi.mock("../api/api");
beforeEach(commonBeforeEach);

// Smoke test
it("renders without crashing", () => {
  render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <PledgedInvestments />
    </UserContext.Provider>
  );
});

// Snapshot test
it("matches snapshot", () => {
  const { asFragment } = render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <PledgedInvestments />
    </UserContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

// Test table rendering
it("renders table with approved requests", async () => {
  render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <PledgedInvestments />
    </UserContext.Provider>
  );

  // Wait for api call to resolve and populate values
  await waitFor(() => {
    expect(BorrowcapApi.getApprovedRequestsByUserId).toHaveBeenCalledWith(
      testUser.id
    );
  });

  // Verify the table is populated
  expect(screen.getByText("Approved Amount")).toBeInTheDocument();
  expect(screen.getByText("Pledged Amount")).toBeInTheDocument();
  expect(screen.getByText("Approved Date")).toBeInTheDocument();
  expect(screen.getByText("Funding Deadline")).toBeInTheDocument();
  expect(screen.getByText("Interest Rate")).toBeInTheDocument();
  expect(screen.getByText("Term")).toBeInTheDocument();

  // Verify the table has correct data
  expect(screen.getByText("$25,000.00")).toBeInTheDocument();
  expect(screen.getByText("$5,000.00")).toBeInTheDocument();
  expect(screen.getByText("2023-01-01")).toBeInTheDocument();
  expect(screen.getByText("2023-02-01")).toBeInTheDocument();
  expect(screen.getByText("7.00%")).toBeInTheDocument();
  expect(screen.getByText("36")).toBeInTheDocument();
});
