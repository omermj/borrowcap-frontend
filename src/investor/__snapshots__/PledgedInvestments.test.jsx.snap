// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="border mb-4"
  >
    <div
      class="py-2"
    >
      <span
        class="align-middle h5"
      >
        Pledged Investments
      </span>
    </div>
    <div
      class=" mb-2 fst-italic fw-light"
    >
       
      <small>
        No pledged investments.
      </small>
    </div>
  </div>
</DocumentFragment>
`;
